trajopt_optimizers/trajopt_sqp/src/qp_problem.cpp

```diff
@@ -265,42 +302,82 @@ void QPProblem::updateSlackVariableBounds()
 
 double QPProblem::evaluateTotalConvexCost(const Eigen::Ref<const Eigen::VectorXd>& var_vals)
 {
-  double result_quad = var_vals.transpose() * hessian_ * var_vals;
-  double result_lin = gradient_.transpose() * var_vals;
-  return result_quad + result_lin;
+  return evaluateConvexCosts(var_vals).sum();
 }
 
-Eigen::VectorXd QPProblem::evaluateConvexCosts(const Eigen::Ref<const Eigen::VectorXd>& /*var_vals*/)
+Eigen::VectorXd QPProblem::evaluateConvexCosts(const Eigen::Ref<const Eigen::VectorXd>& var_vals)
 {
-  return Eigen::VectorXd();
+  if (num_nlp_costs_ == 0)
+    return Eigen::VectorXd();
+
+  auto var_block = var_vals.head(num_nlp_vars_);
+  Eigen::VectorXd result_quad = var_block.transpose() * hessian_.block(0, 0, num_nlp_vars_, num_nlp_vars_) * var_block;
+  Eigen::VectorXd result_lin = var_block.transpose() * gradient_.block(0, 0, num_nlp_vars_, num_nlp_costs_);
+  return cost_constant_ + result_lin + result_quad;
 }
 
-Eigen::VectorXd QPProblem::getExactConstraintViolations()
+Eigen::VectorXd QPProblem::calcConstraintViolations(const Eigen::Ref<const Eigen::VectorXd>& cnt_vals)
 {
-  Eigen::VectorXd cnt_eval = nlp_->EvaluateConstraints(nlp_->GetOptVariables()->GetValues().data());
-
-  // Convert constraint bounds to VectorXd
-  Eigen::VectorXd cnt_bound_lower(num_nlp_cnts_);
-  Eigen::VectorXd cnt_bound_upper(num_nlp_cnts_);
   std::vector<ifopt::Bounds> cnt_bounds = nlp_->GetBoundsOnConstraints();
-  for (Eigen::Index i = 0; i < num_nlp_cnts_; i++)
+  Eigen::VectorXd violation(num_nlp_cnts_);
+  for (Eigen::Index i = 0; i < num_nlp_cnts_; ++i)
   {
-    cnt_bound_lower[i] = cnt_bounds[static_cast<std::size_t>(i)].lower_;
-    cnt_bound_upper[i] = cnt_bounds[static_cast<std::size_t>(i)].upper_;
+    if (constraint_types_[static_cast<std::size_t>(i)] == ConstraintType::EQ)
+    {
+      // The lower and upper should be the same because it is an Equality constraint
+      violation[i] = std::abs(cnt_vals[i] - cnt_bounds[static_cast<std::size_t>(i)].lower_);
+    }
+    else  // ConstraintType::INEQ
+    {
+      // For inequality constraint the valid range is between [0, -inf] so anything greater than
+      // zero is a violation
+      violation[i] = std::max<double>(cnt_vals[i] - cnt_bounds[static_cast<std::size_t>(i)].upper_, 0);
```

Comment:
Does this work if the bounds are not 0 to inf? Before, it was checking to see if it was outside either bound

Reply:
I believe this was one restriction in the original trajopt that constraint had to be written such that they obeyed this but not sure if we want to impose the same restriction. Well, not that I think about it how did we handle tolerances in the old trajopt?

Reply:
I plan to revert to what you had originally. This was mostly to help me track down other issues but what you had originally is what we want.

Reply:
This may also affect the slack variables.

Reply:
This has been resolved with the addition of the `calcBoundsErrors` and `calcBoundsViolations`.